### PR TITLE
Metacontroller monitoring

### PR DESCRIPTION
**Pull request by natalian98**
- Added integration with prometheus and grafana
- Updated requirements with `lightkube-models == 1.21.0.4` due to changes impacting the charm in v1.23

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -37,10 +41,34 @@ def __init__(self, *args):
...
+        )
+
+        self.dashboard_provider = GrafanaDashboardProvider(self)
+
         self.framework.observe(self.on.install, self._install)
         self.framework.observe(self.on.remove, self._remove)
         self.framework.observe(self.on.update_status, self._update_status)
 
+        monitoring_events = [
+            self.on["metrics-endpoint"].relation_changed,
```
Comment (simskij) [inline src/charm.py:64]: Add `relation_joined`

**Review by simskij — CHANGES_REQUESTED** [review]
Good job!

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -46,6 +48,45 @@ async def test_build_and_deploy_with_trust(ops_test: OpsTest):
...
+    logger.info(f"Response status is {response_status}")
+
+    response_metric = response["data"]["result"][0]["metric"]
+    assert response_metric["juju_application"] == APP_NAME
+    assert response_metric["juju_model"] == ops_test.model_name
+
+
 # TODO: Add test for charm removal
 # TODO: Add test that USES metacontroller for something (act on a namespace
 #  given particular metadata, similar to kfp?)
```
Comment (simskij) [inline tests/integration/test_charm.py:92]:
```suggestion
# TODO: Add test for charm removal
# TODO: Add test that USES metacontroller for something (act on a namespace
#  given particular metadata, similar to kfp?)
# TODO: Add test that makes sure the Grafana relation actually works.
```

Does not currently verify the Grafana integration.

File: src/charm.py (hunk 1 above)
Comment (mmanciop) [reply src/charm.py:64]: I also believe we have a constant that can be imported from the library itself. @simskij IIRC we even have an event one can register to, no?

**Review by mmanciop — COMMENTED** [review]
Really good stuff :-)

File: src/charm.py (hunk 1 above)
Comment (mmanciop) [reply src/charm.py:64]: Actually… I do not understand what is the benefit of registering these hooks. @natalian98

**Review by mmanciop — COMMENTED** [review]
I am not sure why there are hook registrations for the metrics endpoint relation, but the rest looks really good

**Review by mmanciop — APPROVED** [review]

**Review by mmanciop — APPROVED** [review]

**Pull request merged**